pkg/controller/deployment/sync.go

```diff
@@ -318,35 +318,38 @@ func (dc *DeploymentController) getNewReplicaSet(d *extensions.Deployment, rsLis
 	createdRS, err := dc.client.ExtensionsV1beta1().ReplicaSets(d.Namespace).Create(&newRS)
 	switch {
 	// We may end up hitting this due to a slow cache or a fast resync of the Deployment.
-	// Fetch a copy of the ReplicaSet. If its PodTemplateSpec is semantically deep equal
-	// with the PodTemplateSpec of the Deployment, then that is our new ReplicaSet. Otherwise,
-	// this is a hash collision and we need to increment the collisionCount field in the
-	// status of the Deployment and try the creation again.
+	// Fetch a copy of the ReplicaSet. If the Deployment owns the ReplicaSet and the
+	// ReplicaSet's PodTemplateSpec matches that of the Deployment, it is our new ReplicaSet.
+	// Otherwise, this is a hash collision and we need to increment the collisionCount field
+	// in the status of the Deployment and try the creation again.
 	case errors.IsAlreadyExists(err):
 		alreadyExists = true
 		rs, rsErr := dc.rsLister.ReplicaSets(newRS.Namespace).Get(newRS.Name)
 		if rsErr != nil {
 			return nil, rsErr
 		}
+
+		controllerRef := metav1.GetControllerOf(rs)
+		if controllerRef != nil && controllerRef.UID == d.UID && deploymentutil.EqualIgnoreHash(&d.Spec.Template, &rs.Spec.Template) {
+			createdRS = rs
+			err = nil
+			break
+		}
```

Comment:
Add a comment for this code block

Reply:
That's what I did before till I realized the comment from line 321 to 324 above has explained about the code block clearly.

Reply:
Should we move down the comment from above `case errors.IsAlreadyExists(err):`? That's the one that explains why we do this.

Reply:
Yes. After inspecting the flow, the comment after `...a fast resync of the Deployment.` should be positioned within the case.